src/foam/box/network/TCPSocketMgr.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.box.network',
+  name: 'TCPSocketMgr',
+
+  javaImports: [
+    "java.net.Socket",
+    "foam.core.X"
+  ],
+
+  properties: [
+    {
+      class: 'Long',
+      name: 'timeout',
+      value: 5000
+    },
+    {
+      class: 'Map',
+      name: 'connectionMap',
+      javaFactory: `
+        return java.util.Collections.synchronizedMap(new java.util.HashMap());
+      `
+    }
+  ],
+
+  methods: [
+    {
+      name: 'add',
+      args: [
+        {
+          name: 'con',
+          type: 'Object'
+        }
+      ],
+      javaCode: `
+        ConnectionBox connectionBox = (ConnectionBox) con;
+        getConnectionMap().put(makeKey(connectionBox.getHost(), connectionBox.getPort()), connectionBox);
+      `
+    },
+    {
+      name: 'remove',
+      args: [
+        {
+          name: 'con',
+          type: 'Object'
+        }
+      ],
+      javaCode: `
+        ConnectionBox connectionBox = (ConnectionBox) con;
+        getConnectionMap().remove(makeKey(connectionBox.getHost(), connectionBox.getPort()));
+      `
+    },
+    {
+      name: 'makeKey',
+      type: 'String',
+      args: [
+        {
+          name: 'host',
+          type: 'String'
+        },
+        {
+          name: 'port',
+          type: 'int'
+        }
+      ],
+      javaCode: `
+        return host + ":" + port;
+      `
+    },
+    {
+      name: 'get',
+      type: 'ConnectionBox',
+      args: [
+        {
+          name: 'x',
+          type: 'X'
+        },
+        {
+          name: 'host',
+          type: 'String'
+        },
+        {
+          name: 'port',
+          type: 'int'
+        }
+      ],
+      javaCode: `
+        ConnectionBox conBox = (ConnectionBox) getConnectionMap().get(makeKey(host, port));
+        if ( conBox != null ) return conBox;
+
+        try {
+          Socket socket = new Socket(host, port);
+          conBox = new ConnectionBox(getX(), socket, host, port);
+          conBox.start();
+          return conBox;
+        } catch ( Exception e ) {
+
+        }
+        return null;
+      `
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(foam.java.Code.create({
+          data:`
+            private static TCPSocketMgr instance__;
```

Comment:
This can be replaced with a NSpec script with a serviceClass of TPCSocketMgr, then find the TCPSocketMgr via x.get("tcpSocketMgr").